cpp/tests/streaming/test_shuffler.cpp

```diff
@@ -226,41 +201,263 @@ std::pair<Node, Node> shuffler_nb(
     // extract task: extract the packed chunks from the shuffler and send them to the
     // output channel
     auto extract_task = [](auto shuffler_ctx, auto ctx, auto ch_out) -> Node {
-        ShutdownAtExit c{ch_out};
+        ShutdownAtExit c{
+            ch_out
+        };  // TODO: could this be problematic with multiple consumers?
         co_await ctx->executor()->schedule();
 
-        // wait for all partitions to be finished
-        {
-            auto lock = co_await shuffler_ctx->mtx.scoped_lock();
-            co_await shuffler_ctx->cv.wait(lock, [&shuffler_ctx]() -> bool {
-                return shuffler_ctx->finished;
-            });
-        }
+        while (!shuffler_ctx->shuffler->finished() || !shuffler_ctx->ready_pids.empty()) {
+            auto expected = co_await shuffler_ctx->ready_pids.pop();
+            if (!expected.has_value()) {  // queue is shutdown, so exit the loop
+                break;
+            }
+
+            auto packed_chunks = shuffler_ctx->shuffler->extract(*expected);
+
+            co_await ch_out->send(
+                std::make_unique<PartitionVectorChunk>(
+                    *expected, std::move(packed_chunks)
+                )
+            );
 
+            if (shuffler_ctx->shuffler->finished()) {
+                // if the shuffler is finished, shutdown & drain the ready_pids queue
+                co_await shuffler_ctx->ready_pids.shutdown_drain(ctx->executor());
+            }
+        }
         co_await ch_out->drain(ctx->executor());
     };
 
-    return {
-        insert_task(shuffler_ctx, ctx, total_num_partitions, stream, std::move(ch_in)),
+    std::vector<Node> nodes;
+    nodes.emplace_back(
+        insert_task(shuffler_ctx, ctx, total_num_partitions, stream, std::move(ch_in))
+    );
+    for (int i = 0; i < n_consumers - 1; ++i) {
+        nodes.emplace_back(extract_task(shuffler_ctx, ctx, ch_out));
+    }
+    nodes.emplace_back(
         extract_task(std::move(shuffler_ctx), std::move(ctx), std::move(ch_out))
-    };
+    );
+
+    co_await coro::when_all(std::move(nodes));
 }
 
 }  // namespace
 
-TEST_F(StreamingShuffler, callbacks) {
-    EXPECT_NO_FATAL_FAILURE(
-        run_test([&](auto ctx, auto ch_in, auto ch_out, std::vector<Node>& nodes) {
-            auto [insert_node, extract_node] = shuffler_nb(
-                std::move(ctx),
-                stream,
-                std::move(ch_in),
-                std::move(ch_out),
-                op_id,
-                num_partitions
+TEST_P(StreamingShuffler, callbacks_1_consumer) {
+    EXPECT_NO_FATAL_FAILURE(run_test([&](auto ch_in, auto ch_out) -> Node {
+        return shuffler_nb(
+            ctx, stream, std::move(ch_in), std::move(ch_out), op_id, num_partitions, 1
+        );
+    }));
+}
+
+TEST_P(StreamingShuffler, callbacks_2_consumer) {
+    GTEST_SKIP() << "unreliable test";  // TODO: fix this
```

Comment:
What is unreliable about this?

Reply:
Multiple consumers were unreliable when I pin the tests to a single core. It was deadlocking intermittently. 
This will be fixed in https://github.com/rapidsai/rapidsmpf/pull/516